Ex15 js oop

DarkProDen opened the pull request.

DarkProDen committed:
add classwork

DarkProDen committed:
Add Task15

eugrdn commented on src/ex15_js-oop/Classwork.js:
```diff
@@ -0,0 +1,55 @@
+    this.a = a;
+}
+Square.prototype = shape;
+Square.prototype.getPerimeter = function () {
```
не сохранил конструктор, и ниже тоже

DarkProDen replied:
Исправил (вроди)

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,182 @@
+}
+
+var Lamp = function (name, power) {
+    this.name = name;
```
```
this.name = ...
this.type = ...
this.power = ...
this.enabled = ...
```
повторяется во всех конструкторах: переиспользуй родительский

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,182 @@
+    this.luminousFlux = 2000;
+    this.lamps = [new Lamp(name+'__lamp1', power / 3), new Lamp(name+'__lamp2', power / 3), new Lamp(name+'__lamp3', power / 3)];
+}
+Chandelier.prototype = new Device();
```
так лучше не делать, прочитай https://learn.javascript.ru/class-inheritance#polnyy-kod-nasledovaniya про неправильный вариант

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,182 @@
+}
+Chandelier.prototype = new Device();
+Chandelier.prototype.switchOnOff = function () {
+    for (let i = 0; i < Object(this.lamps).length; i++) {
```
зачем обертка `Object(...)`? и ниже

DarkProDen replied:
я пытался сначала this.lamps.length . выдает ошибку.

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,182 @@
+    this.name = name;
+    this.devices = devices;
+    /*
+    devices.forEach(device => {
```
коммент?

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,182 @@
+    new Device('Microwave', 2000),
+    new Screen('TV', 300)]);
+var i=0;
+room.devices.forEach(device => {
```
2 аргумент колбека -- index, не нужно создать глобальную переменную

DarkProDen committed:
Update Classwork.js

Constructors saved.

DarkProDen committed:
Update Task15.js

Changes:
Device.prototype.constructor = Device;
Device.apply(this, arguments);
Refrigerator.prototype = Object.create(Device.prototype);

DarkProDen commented:
Исправил все

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,170 @@
+Chandelier.prototype = Object.create(Device.prototype);
+Chandelier.prototype.constructor = Chandelier;
+Chandelier.prototype.switchOnOff = function () {
+    for (let i = 0; i < this.lamps.length; i++) {
```
используй forEach

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,170 @@
+
+var Room = function (name, devices) {
+    this.name = name;
+    this.devices = [];
```
а почему не просто `this.devices = devices`?

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,170 @@
+Room.prototype.constructor = Room;
+Room.prototype.getPower = function () {
+    var result = 0;
+    for (let i = 0; i < this.devices.length; i++) {
```
используй reduce

eugrdn commented on src/ex15_js-oop/Task15.js:
```diff
@@ -0,0 +1,170 @@
+    return result;
+}
+Room.prototype.getDevice = function (deviceName) {
+    for (let i = 0; i < this.devices.length; i++) {
```
используй find

DarkProDen replied:
Сделал все

DarkProDen committed:
Update Task15.js

Used foreach(), find(), reduce();
this.devices = devices.
